---
title: "Rsession_02_filter_genotypes_and_variants"
author: "Mae Goldgraben"
date: "28May2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

last updated: Mae Goldgraben, 28May2018  

# Notes 

Source data includes - ihcap_75_Aug17_vqsr_D0Q100 - annotated vcf which includes 3 more samples for vqsr model. 

The hard filters will be applied in the following order:  
gq > 20  
dp < 10x median coverage (median is 106) 
dp >= 10 reads per sample  
call_rate > 0.5 (as uniform cohort)  

Filters consistent with filtering applied to wecare-1K  
- The gq filter >20 is arbitrary; however, consistent with what is done by others  
(e.g. see Carson BMC Bioinformatics. 2014 15:125).  

A small number of genotypes is covered by too high or too low sequencing depth to be true.  
These are obvious mistakes, and they have been removed too by choosing an arbitrary threshold   
for max DP of 10x the median sequencing coverage and min DP at 10 reads.  

The call rate filter is at 50%, as this is a uniform library dataset.

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Clean up
rm(list=ls())

# Load libraries
library(knitr)
library(dplyr)
library(stringr)
library(VennDiagram)

# Folders
setwd("/rds/project/erf33/rds-erf33-medgen/users/mae/IHCAP75_atm_aml/germline_analysis_all_timepts/Rscripts_and_filters")
source_data_folder <- "/rds/project/erf33/rds-erf33-medgen/users/mae/IHCAP75_atm_aml/germline_analysis_all_timepts/Rscripts_and_filters/source_data"
interim_data_folder <- "/rds/project/erf33/rds-erf33-medgen/users/mae/IHCAP75_atm_aml/germline_analysis_all_timepts/Rscripts_and_filters/interim_data"

# Thresholds for genotype quality score & depth
min.gq <- 20
min.dp <- 10
max.dp <- 1060

# Variants call rate threshold
min_call_rate <- 0.5

```

# load_data

```{r load_data}

load(paste(interim_data_folder, "Rsession_01_read_data_germ.RData", sep="/"))

```

# check_data

```{r check_data}

dim(gt.mx)
class(gt.mx)
gt.mx[1:5,1:5]

dim(gq.mx)
class(gq.mx)
gq.mx[1:5,1:5]

dim(dp.mx)
class(dp.mx)
dp.mx[1:5,1:5]

dim(vv.df)
str(vv.df)
vv.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

# Check consistence of rownames
sum(rownames(gt.mx) != rownames(gq.mx))
sum(rownames(gt.mx) != rownames(dp.mx))
sum(rownames(gt.mx) != rownames(vv.df))
sum(rownames(gt.mx) != rownames(kgen.df))
sum(rownames(gt.mx) != rownames(exac.df))

# Check consistence of colnames
sum(colnames(gt.mx) != colnames(gq.mx))
sum(colnames(gt.mx) != colnames(dp.mx))

```

# explore_data_before_filtering

Calculate genotype NA rates
Histogram of call rates per variant, gq and dp in non-NA genotypes 

```{r explore_data_before_filtering}

# Fraction of NA genotypes before filtering
sum(is.na(gt.mx))/(nrow(gt.mx)*ncol(gt.mx)) # ~1.4%; 
# Was 4% for wecare

# Call rates per variant before filtering
x <- ncol(gt.mx)
y <- apply(gt.mx,1,function(z){1-sum(is.na(z))/x})
hist(y, breaks=100,xlab="Call rate", ylab="Number of Variants",
     main="Call rates per variant before genotype filtering")

# Histogram of gq before filtering (when gt is not NA !)
hist(gq.mx[!is.na(gt.mx)], breaks=50, xlab="Genotype Quality", ylab="Number of Variants", 
     main="Histogram of gq in non-NA genotypes (before filtering)")

# Histogram of dp before filtering (when gt is not NA !)
hist(dp.mx[!is.na(gt.mx)], breaks=50, xlab=NULL, main="Histogram of dp in non-NA genotypes (before filtering)")
hist(dp.mx[!is.na(gt.mx)], breaks=2500, xlab = "Depth", xlim=c(0,500), main="Histogram of dp in non-NA genotypes (before filtering, 0:500)")

# Clean-up
rm(x,y)

```

# filter_out_low_gq

Put NA to genotypes where gq < 20 : removes ~5% of non-NA genotypes

```{r filter_out_low_gq}

# num of genotypes that are NA and number to be removed from gt.mx
sum(gq.mx < min.gq, na.rm=TRUE) # 24,861
sum(gq.mx[!is.na(gt.mx)] < min.gq, na.rm=TRUE) # 24,861

# Fraction of genotypes to be removed
sum(!is.na(gt.mx)) # 464,530
sum(gq.mx < min.gq, na.rm=TRUE)/sum(!is.na(gq.mx)) # ~5% of gq.mx ; wecare was 8%
sum(gq.mx[!is.na(gt.mx)] < min.gq, na.rm=TRUE) / sum(!is.na(gt.mx)) # ~5% of gt.mx

# Apply filter (to gt only !)
NA -> gt.mx[ gq.mx < min.gq ]

# Clean up
rm(min.gq)

```

# explore_data_after_gq_filtering

Calculate genotype NA rates
Histogram of call rates per variant, gq and dp in non-NA genotypes

```{r explore_data_after_gq_filtering}

dim(gt.mx)

# Fraction of NA genotypes after gq filtering
sum(is.na(gt.mx))/(dim(gt.mx)[1]*dim(gt.mx)[2]) # ~6%

# Call rates per variant after gq filtering
x <- ncol(gt.mx)
y <- apply(gt.mx, 1, function(z){1-sum(is.na(z))/x})
hist(y, breaks=100, xlab="Call rate", ylab="Number of Variants",
     main="Histogram of call rates per variant after gq filtering")

# Histogram of gq after gq filtering (when gt is not NA !)
hist(gq.mx[!is.na(gt.mx)], xlim=c(0,100), xlab=NULL, breaks=50, main="Histogram of gq in non NA genotypes (after gq filtering)")

# Histogram of dp after gt filtering (when gt is not NA !)
hist(dp.mx[!is.na(gt.mx)], breaks=50, xlab="Read depth", main="Histogram of dp in non-NA genotypes (after gq filtering)")
hist(dp.mx[!is.na(gt.mx)], xlim=c(0,500), xlab="Read depth", breaks=2500, main="Histogram of dp in non-NA genotypes (after gq filtering, 0:500)")

# Clean up
rm(x, y)

```

# filter_out_high_dp

put NA to genotypes where dp > 10x median coverage (i.e. 1060) : removes ~<1% of non-NA genotypes

```{r filter_out_high_dp}

# Calculate number of max DP variants to be removed from gt.mx
dp_max.mx <- as.matrix(dp.mx > max.dp)
sum(dp_max.mx, na.rm=TRUE) # 1
sum(is.na(dp_max.mx)) # 2
sum(dp_max.mx[!is.na(gt.mx)], na.rm=TRUE) # 1

# Fraction of genotypes to be removed (approx)
sum(!is.na(gt.mx)) # 439,669
sum(dp_max.mx[!is.na(gt.mx)], na.rm=TRUE) / sum(!is.na(gt.mx)) # 0.000002; <1%

# Apply filter (to gt only!; can only run on matrices)
NA -> gt.mx[dp_max.mx]
sum(!is.na(gt.mx)) # 439,668

# Clean up
rm(dp_max.mx, max.dp)

```

# explore_data_after_gq_Hdp_filtering

Calculate genotype NA rates 
Histogram of call rates per variant, gq and dp  

```{r explore_data_after_gq_Hdp_filtering}

dim(gt.mx)

# Fraction of NA genotypes after gq + Hdp filtering
sum(is.na(gt.mx))/(dim(gt.mx)[1]*dim(gt.mx)[2]) # 6.67%

# Call rates per variant after gq + Hdp filtering
x <- ncol(gt.mx)
y <- apply(gt.mx, 1, function(z){1-sum(is.na(z))/x})
hist(y, breaks=100, xlab="Call rate", ylab="Number of Variants", main="Histogram of call rates per variant after gq+Hdp filtering")

# Histogram of dp after gt+Hdp filtering (when gt is not NA !)
hist(dp.mx[!is.na(gt.mx)], xlab="Read depth", breaks=2500, main="Histogram of dp after gt+Hdp filtering")
hist(dp.mx[!is.na(gt.mx)], xlim=c(0,500), xlab="Read depth", breaks=2500, main="Histogram of dp in non-NA genotypes (after gt+Hdp filtering, 0:500)")
hist(dp.mx[!is.na(gt.mx)], xlim=c(0,100), xlab="Read depth", breaks=1000, main="Histogram of dp in non-NA genotypes (after gt+Hdp filtering, 0:100)")

# Clean up
rm(x, y)

```

# filter_out_low_dp

put NA to genotypes where dp < 10 per sample, removes 8% of variants

```{r filter_out_low_dp}

# Num of genotypes that are NA and number to be removed from gt.mx
sum(dp.mx < min.dp, na.rm=TRUE) # 37,899
sum(dp.mx[!is.na(gt.mx)] < min.dp, na.rm=TRUE) # 12,598

# Fraction of genotypes to be removed (approx)
sum(!is.na(gt.mx)) # 439,668
sum(dp.mx < min.dp, na.rm=TRUE)/sum(!is.na(dp.mx)) # ~8% of dp.mx
sum(dp.mx[!is.na(gt.mx)] < min.dp, na.rm=TRUE) / sum(!is.na(gt.mx)) # 0.028; <3% of gt.mx

# Apply filter (to gt only!; can only run on matrices)
NA -> gt.mx[ dp.mx < min.dp ]
sum(!is.na(gt.mx)) # 427,070

# Clean up
rm(min.dp)

```

# explore_data_after_gq_Hdp_Ldp_filtering

Calculate genotype NA rates 
Histogram of call rates per variant and dp

```{r explore_data_after_gq_Hdp_Ldp_filtering}

dim(gt.mx)

# Fraction of NA genotypes after gq + Hdp + Ldp filtering
sum(is.na(gt.mx))/(dim(gt.mx)[1]*dim(gt.mx)[2]) # 9%

# Call rates per variant after gq + Hdp + Ldp filtering
x <- ncol(gt.mx)
y <- apply(gt.mx, 1, function(z){1-sum(is.na(z))/x})
hist(y, breaks=100, xlab="Call rate", ylab="Number of Variants", 
     main="Histogram of call rates per variant after gq+Hdp+Ldp filtering")

# Histogram of dp after gt+Hdp+Ldp filtering (when gt is not NA !)
hist(dp.mx[!is.na(gt.mx)], xlab="Read depth", breaks=2500, main="Histogram of dp after gt+Hdp+Ldp filtering")
hist(dp.mx[!is.na(gt.mx)], xlim=c(0,100), xlab="Read depth", breaks=1000, 
     main="Histogram of dp in non-NA genotypes \n (after gt+Hdp+Ldp filtering, 0:100)")

# Clean up
rm(x, y)

```

# filter_variants_by_cohort_call_rate

Remove variants with call rate < 50%

```{r filter_variants_by_cohort_call_rate}

# Look at all non-NAs in gt.mx
dim(gt.mx) # 58,884 x 8 = 471,072
sum(!is.na(gt.mx))/(nrow(gt.mx)*ncol(gt.mx)) # 90% of in gt.mx variants are good

# Calculate the "good call rate variants" & estimate the proportion of variants to be retained
call_rate.udf <- function(x){sum(!is.na(x))/length(x)}

call_rate <- apply(gt.mx, 1, call_rate.udf)
hist(call_rate)
good_variants <- call_rate >= min_call_rate
summary(good_variants) # 55,579
sum(good_variants)/nrow(gt.mx) # 94% variants retained

# Apply call rate filter to gt.mx
sum(is.na(gt.mx)) # 44,002
NA -> gt.mx[!good_variants,]
sum(is.na(gt.mx)) # 49,210
dim(gt.mx) # 58,884 x 8

# Clean-up
rm(min_call_rate, call_rate.udf, call_rate, good_variants)

```

# cleanup_good_call_rate_gt_mxs

Remove poor quality variants across all subsets, clean up vv/kgen/exac files

```{r cleanup_good_call_rate_gt_mxs}

# Clean up gt matrix for very poor variants
count_nas.udf <- function(x){sum(is.na(x))}

num_of_nas <- apply(gt.mx, 1, count_nas.udf)
hist(num_of_nas)
very_poor_variants <- num_of_nas == ncol(gt.mx)
summary(very_poor_variants) # 3,305

dim(gt.mx) # 58,884 x 8
cleaned_gt.mx <- gt.mx[!very_poor_variants,]
dim(cleaned_gt.mx) # 55,579 x 8

# ---- Clean up vv.df, exac.df and kgen.df ---- #

cleaned_variants <- rownames(cleaned_gt.mx)
  
cleaned_vv.df <- vv.df[cleaned_variants, ]
dim(cleaned_vv.df) # 55,579 x 68
cleaned_kgen.df <- kgen.df[cleaned_variants, ]
dim(cleaned_kgen.df) # 55,579 x 9
cleaned_exac.df <- exac.df[cleaned_variants, ]
dim(cleaned_exac.df) # 55,579 x 48

# Clean-up
rm(count_nas.udf, num_of_nas, very_poor_variants, cleaned_variants)

```

# check_remaining_files_and_remove_unnecessary_data

```{r check_remaining_files_and_remove_unnecessary_data}

# Check consistency of rownames
sum(rownames(cleaned_gt.mx) != rownames(cleaned_vv.df))
sum(rownames(cleaned_gt.mx) != rownames(cleaned_kgen.df))
sum(rownames(cleaned_gt.mx) != rownames(cleaned_exac.df))

# Clean-up

rm(gt.mx, gq.mx, dp.mx, vv.df, kgen.df, exac.df)

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "Rsession_02_filter_genotypes_and_variants_germ.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```

